bundler/lib/bundler/self_manager.rb

```diff
@@ -26,15 +26,27 @@ def install_locked_bundler_and_restart_with_it_if_needed
 
     def install_and_restart_with_locked_bundler
       bundler_dep = Gem::Dependency.new("bundler", lockfile_version)
+      spec = fetch_spec_for(bundler_dep)
+      unless spec
+        Bundler.ui.warn "Your lockfile is locked to a version of bundler (#{lockfile_version}) that doesn't exist at https://rubygems.org."
+        return
+      end
 
-      Gem.install(bundler_dep)
+      spec.source.install(spec)
     rescue StandardError => e
       Bundler.ui.trace e
       Bundler.ui.warn "There was an error installing the locked bundler version (#{lockfile_version}), rerun with the `--verbose` flag for more details. Going on using bundler #{current_version}."
     else
       restart_with_locked_bundler
     end
 
+    def fetch_spec_for(bundler_dep)
+      source = Bundler::Source::Rubygems.new("remotes" => "https://rubygems.org")
+      source.remote!
```

Comment:
This needs to be done unless the `--local` flag was given to `bundle install`. I'll improve this tomorrow.

Reply:
This issue is already present and there are more pressing things, so I will reticket it.